### PR TITLE
Permit a property to be declared as AD if its old state has already been requested

### LINKED ISSUE #19116: getMaterialPropertyOld marks the property as regular
## Bug Description

I have an object A that retrieves an old material property in its initialization list
```cpp
_strain_old(getMaterialPropertyOldByName<RankTwoTensor>("total_strain"))
```
and an object B that declares an AD material property as
```cpp
_total_strain(declareADProperty<RankTwoTensor>("total_strain"))
```
Then I get an error message like this
```
*** ERROR ***
Attempting to declare AD material property total_strain, but it is already retrieved/declared as a regular material property.
```

## Steps to Reproduce

A MWE [here](https://github.com/hugary1995/raccoon/blob/62c9841c3c73ffa79afa9a5dac0c5f508fc7aa24/tutorials/hydraulic_fracturing_SPF/test.i).

## Impact

This bug leads to errors that are construction order dependent.

### PR DESCRIPTION
**Pull request by dschwen**
Closes #19116

### CODE REVIEW

File: framework/include/materials/MaterialData.h [hunk 1]
```diff
@@ -275,20 +278,28 @@ MaterialData::getPropertyHelper(const std::string & prop_name,
   // Register/get the ID of the property
   const auto prop_id = addPropertyHelper(
       prop_name, typeid(T), state, declare ? &castRequestorToDeclarer(requestor) : nullptr);
+  const auto size = prop_id + 1;
+
+  // get the maximum state id requested for this property so far
+  if (_max_state_requested.size() < size)
+    _max_state_requested.resize(size, 0);
```
Comment (loganharbour) [inline framework/include/materials/MaterialData.h:285]: This data should already exist in `MaterialPropertyStorage`:

```
_storage.getPropRecord(prop_id).state
```

**Review by loganharbour — CHANGES_REQUESTED** [review]
so simple

File: framework/include/materials/MaterialData.h [hunk 2]
```diff
@@ -275,20 +278,28 @@ MaterialData::getPropertyHelper(const std::string & prop_name,
...
     {
-      std::unique_ptr<PropertyValue> value =
-          state_i == 0 ? std::make_unique<GenericMaterialProperty<T, is_ad>>(prop_id)
-                       : _props[0][prop_id].clone(0);
-      entry.setPointer(prop_id, std::move(value), {});
+      if (state == 0)
+        entry.setPointer(
+            prop_id, std::move(std::make_unique<GenericMaterialProperty<T, is_ad>>(prop_id)), {});
+      else
+        entry.setPointer(prop_id, std::move(std::make_unique<MaterialProperty<T>>(prop_id)), {});
```
Comment (loganharbour) [inline framework/include/materials/MaterialData.h:302]: I'm trying to figure out how this now works without a clone... that used to not be the case

File: framework/include/materials/MaterialData.h (hunk 1 above)
Comment (loganharbour) [reply framework/include/materials/MaterialData.h:285]: Unfortunately you'll need to add it to the source file due to includes...

Comment (dschwen) [reply framework/include/materials/MaterialData.h:285]: Circular dependency. Refactoring that will be a pain.

Comment (dschwen) [reply framework/include/materials/MaterialData.h:285]: No, easy. I put a function to get that value in MaterialData.C. That might be what you meant in the first place.

Comment (loganharbour) [reply framework/include/materials/MaterialData.h:285]: Yep, that's what I meant

**Review by loganharbour — DISMISSED** [review]

**Comment by tophmatthews** [comment]
Whoa just ran into this problem, and here's the fix! Yay!

**Comment by lindsayad** [comment]
beautiful change

**Comment by loganharbour** [comment]
@dschwen looks like there's some issues with threads? Trying to think where the disparity would be

**Review by loganharbour — APPROVED** [review]

**Pull request merged**